### PR TITLE
Poor man's cobalt target

### PR DESCRIPTION
**Pull request by jellefoks**
Minimal cobalt target. Works on Linux only (for now).

$ autoninja -C out/chrobalt cobalt
$ out/chrobalt/cobalt

It also enables the devtools listening on 9222.

b/374191454

### CODE REVIEW

File: BUILD.gn [hunk 1]
```diff
@@ -92,6 +92,7 @@ group("gn_all") {
     deps = [
       "//base:base_perftests",
       "//base:base_unittests",
+      "//cobalt",
```
Comment (kaidokert) [inline BUILD.gn:95]: We don't want to modify upstream here without gating it with a build flag - the original Chrome code should keep building without modifications. @andrewsavage1 please advice what `if() clause to add at the bottom here.

File: content/shell/browser/shell_content_browser_client.cc [hunk 2]
```diff
@@ -625,6 +625,14 @@ base::FilePath ShellContentBrowserClient::GetFirstPartySetsDirectory() {
 }
 
 std::string ShellContentBrowserClient::GetUserAgent() {
+  base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
```
Comment (kaidokert) [inline content/shell/browser/shell_content_browser_client.cc:628]: This code should live under /cobalt somewhere, not as a direct modification to upstream. If that's not viable at all, it should be either run-time gated with a `cobalt` conditional ( preferable ) or runtime-gated if #ifdef

File: content/public/common/content_switches.h [hunk 3]
```diff
@@ -224,6 +224,7 @@ CONTENT_EXPORT extern const char kUseFakeUIForFedCM[];
 CONTENT_EXPORT extern const char kUseFakeUIForMediaStream[];
 CONTENT_EXPORT extern const char kVideoImageTextureTarget[];
 CONTENT_EXPORT extern const char kUseMobileUserAgent[];
+CONTENT_EXPORT extern const char kUseCobaltUserAgent[];
```
Comment (kaidokert) [inline content/public/common/content_switches.h:227]: Same comment as elsewhere - this should live under /cobalt, be runtime-gated or as a last resort, #ifdef gated for Cobalt.

File: BUILD.gn (hunk 1 above)
Comment (jellefoks) [reply BUILD.gn:95]: Thanks Andrew. Done.

File: content/shell/browser/shell_content_browser_client.cc (hunk 2 above)
Comment (jellefoks) [reply content/shell/browser/shell_content_browser_client.cc:628]: Ok. I can subclass `ShellMainDelegate` to an implementation that replaces `CreateContentBrowserClient` with one that instantiates a subclass of `ShellContentBrowserClient` that returns a Cobalt string for `GetUserAgent`.

File: content/public/common/content_switches.h (hunk 3 above)
Comment (jellefoks) [reply content/public/common/content_switches.h:227]: I can remove this switch with the change I proposed above. Working on it now.

File: cobalt/BUILD.gn [hunk 4]
```diff
@@ -13,6 +13,53 @@
...
+    testonly = true
+    deps = [ "//content/shell/android:content_shell_apk" ]
+  }
+} else {
+  executable("cobalt") {
+    testonly = true
+
+    sources = [
+      "cobalt.cc",
+      "cobalt_content_browser_client.cc",
```
Comment (johnxwork) [inline cobalt/BUILD.gn:40]: Can we enable these cobalt client implementations for Android too? I think Android can share these?

Comment (jellefoks) [reply cobalt/BUILD.gn:40]: Sure, that would make a lot of sense to do, but this PR doesn't include an Android target.

File: cobalt/cobalt.cc [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  // TODO: Move this logic to ContentMainParams or ContentMainDelegate.
+  static const char* my_argv[] = {
+      argv[0],
+      "--disable-fre",
+      "--no-first-run",
+      "--kiosk",
+      "--force-video-overlays",
+      "--remote-debugging-port=9222",
+      "--remote-allow-origins=http://localhost:9222",
+      "--use-cobalt-user-agent",
```
Comment (borongc) [inline cobalt/cobalt.cc:46]: Please add `--user-level-memory-pressure-signal-params` as it will fail on CHECK() on m114. See b/375060114.

File: cobalt/BUILD.gn [hunk 6]
```diff
@@ -13,6 +13,53 @@
...
   # TODO(b/371589344): Fix android build configs.
-  deps = [ "//starboard($starboard_toolchain)" ]
+  deps = [
+    ":cobalt",
+    "//starboard($starboard_toolchain)",
+  ]
+}
+
+if (is_android) {
+  import("//build/config/android/config.gni")
```
Comment (andrewsavage1) [inline cobalt/BUILD.gn:26]: Why is this import needed?

File: cobalt/cobalt.cc [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+// window to endless possibilities, offering a refreshing breeze on a summer day
+// or a warm embrace on a chilly night. With each click of the remote, a new
+// adventure unfolded, painting a tapestry of personal journeys and shared
+// experiences.
+
+int main(int argc, const char** argv) {
+  cobalt::CobaltMainDelegate delegate;
+  content::ContentMainParams params(&delegate);
+
+  // TODO: Move this logic to ContentMainParams or ContentMainDelegate.
```
Comment (andrewsavage1) [inline cobalt/cobalt.cc:37]: Let's open and add a bug here

File: cobalt/cobalt_content_browser_client.cc [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+#define COBALT_VERSION "26.lts.0-qa"
+
+std::string GetCobaltUserAgent() {
+  return std::string(
+      "Mozilla/5.0 (X11; Linux x86_64) Cobalt/26.lts.0-qa (unlike Gecko) "
+      "v8/unknown gles Starboard/17, "
+      "SystemIntegratorName_DESKTOP_ChipsetModelNumber_2025/FirmwareVersion "
+      "(BrandName, ModelName)");
+
+  // std::string product = COBALT_BRAND_NAME "/" COBALT_VERSION;
```
Comment (andrewsavage1) [inline cobalt/cobalt_content_browser_client.cc:34]: Can we either delete this commented code or use it?

File: cobalt/cobalt.cc [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      "--force-video-overlays",
+      "--remote-debugging-port=9222",
+      "--remote-allow-origins=http://localhost:9222",
+      "--use-cobalt-user-agent",
+      "https://www.youtube.com/tv",
+      nullptr,
+      nullptr};
+  int my_argc = 9;
+
+  if ((!strcmp(argv[0], "/proc/self/exe")) ||
```
Comment (andrewsavage1) [inline cobalt/cobalt.cc:53]: Consider either adding a comment or putting this condition into descriptive functions to make it clear what's happening here—I can see I'm supposed to call this with specific flags, but I can't tell what flags would do what

**Review by andrewsavage1 — COMMENTED** [review]
Looks good outside of the comments, can approve once those get addressed

File: cobalt/cobalt.cc (hunk 5 above)
Comment (jellefoks) [reply cobalt/cobalt.cc:46]: Sure thing. Done.

File: cobalt/BUILD.gn (hunk 6 above)
Comment (jellefoks) [reply cobalt/BUILD.gn:26]: Most of this is taken I left that in from content/shell/BUILD.gn to keep it a little closer for Android to work too. I can strip the entire `if (is_android)` is that is more clear.

File: cobalt/cobalt.cc [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  static const char* my_argv[] = {
+      argv[0],
+      "--disable-fre",
+      "--no-first-run",
+      "--kiosk",
+      "--force-video-overlays",
+      "--remote-debugging-port=9222",
+      "--remote-allow-origins=http://localhost:9222",
+      "--use-cobalt-user-agent",
+      "--user-level-memory-pressure-signal-params",
```
Comment (borongc) [inline cobalt/cobalt.cc:47]: Maybe also put a comment above this flag `--user-level-memory-pressure-signal-params`, as this is specific to m114 and we should be removed it after we rebase to m120+ in the future.

File: cobalt/cobalt.cc (hunk 7 above)
Comment (jellefoks) [reply cobalt/cobalt.cc:37]: Done, b/375241103 (PR update follows).

File: cobalt/cobalt.cc (hunk 9 above)
Comment (jellefoks) [reply cobalt/cobalt.cc:53]: Done, PR update follows.

File: cobalt/cobalt_content_browser_client.cc (hunk 8 above)
Comment (jellefoks) [reply cobalt/cobalt_content_browser_client.cc:34]: Deleted, and bug added for not hard-coding the UA.

File: cobalt/BUILD.gn (hunk 6 above)
Comment (jellefoks) [reply cobalt/BUILD.gn:26]: Deleted. PR update follows

**Review by kaidokert — APPROVED** [review]
LGTM

File: cobalt/BUILD.gn [hunk 11]
```diff
@@ -13,6 +13,44 @@
 # limitations under the License.
 
 group("gn_all") {
+  testonly = true
```
Comment (yell0wd0g) [inline cobalt/BUILD.gn:16]: GN `testonly` is reserved for targets that produce test files.
So this should be removed. Same with l.27. See

https://source.chromium.org/search?q=file:BUILD.gn%20testonly%20file:(content%7Cbase%7Cmedia)&ss=chromium%2Fchromium%2Fsrc

for some uses.

File: cobalt/BUILD.gn [hunk 12]
```diff
@@ -13,6 +13,44 @@
...
+
   # TODO(b/371589344): Fix android build configs.
   deps = [ "//starboard($starboard_toolchain)" ]
+  if (!is_android) {
+    deps += [ ":cobalt" ]
+  }
+}
+
+if (!is_android) {
+  executable("cobalt") {
```
Comment (yell0wd0g) [inline cobalt/BUILD.gn:26]: GN-style I'd rather formulate this like:
```
executable("cobalt") {
  // TODO(...): Support Android TV (target_os == android),
  assert(!is_android);
```

to reduce the indentation.

File: cobalt/BUILD.gn [hunk 13]
```diff
@@ -13,6 +13,44 @@
...
+      "//content/shell:pak",
+      "//sandbox",
+    ]
+
+    data_deps = [
+      "//content/shell:pak",
+      "//tools/v8_context_snapshot:v8_context_snapshot",
+    ]
+
+    data_deps += [ "//components/crash/core/app:chrome_crashpad_handler" ]
```
Comment (yell0wd0g) [inline cobalt/BUILD.gn:52]: I'd suggest removing this because a) if we crash, we crash locally, so
there's no point in having a handler to submit to a server and b) if
we send crash reports to a server, they'll be stored as if they were
Chrome M114s and skew other folks data : )

File: cobalt/cobalt.cc [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  content::ContentMainParams params(&delegate);
+
+  // TODO: (cobalt b/375241103) Reimplement this in a clean way.
+  // This defines a list of command-line overrides. When adding or removing
+  // parameters, also update the my_argc below.
+  static const char* my_argv[] = {
+      argv[0], "--disable-fre", "--no-first-run", "--kiosk",
+      "--force-video-overlays",
+      // Enable remote devtools access.
+      "--remote-debugging-port=9222",
```
Comment (yell0wd0g) [inline cobalt/cobalt.cc:44]: Isn't this the default value?

File: cobalt/cobalt.cc [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+      "--remote-allow-origins=http://localhost:9222",
+      // This flag is added specifically for m114 and should be removed after
```
Comment (yell0wd0g) [inline cobalt/cobalt.cc:46]: If this is a TODO then it should be marked so:  `//TODO(b/...): Reenable...`.
If it's just an FYI, then leave a link to the bug b/375060114 .

File: cobalt/cobalt.cc [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+      // rebasing to m120+
+      "--user-level-memory-pressure-signal-params",
+      "https://www.youtube.com/tv", nullptr, nullptr};
+  int my_argc = 9;
+
+  // TODO: (cobalt b/375241103) Reimplement this in a clean way.
```
Comment (yell0wd0g) [inline cobalt/cobalt.cc:52]: nit: s/TODO: (cobalt b/375241103)/TODO(b/375241103):/

Or just TODO(375241103) since now Chrome CS understands
Buganizer numbers.

**Review by yell0wd0g — COMMENTED** [review]
Just have a few nits post-commit.

File: cobalt/BUILD.gn (hunk 11 above)
Comment (jellefoks) [reply cobalt/BUILD.gn:16]: If i remove it, it won't build because the top level gn_all has it.

File: cobalt/BUILD.gn (hunk 12 above)
Comment (jellefoks) [reply cobalt/BUILD.gn:26]: This is left under an if, to match content/shell/BUILD.gn, as for Android it's not an "executable" but a "group", and that will have to be brought in here.

File: cobalt/cobalt.cc (hunk 15 above)
Comment (jellefoks) [reply cobalt/cobalt.cc:46]: I followed the format proposed and not contradicted in the thread.

File: cobalt/BUILD.gn (hunk 13 above)
Comment (jellefoks) [reply cobalt/BUILD.gn:52]: Good point. This was just taken to match content/shell without giving it any thought. I'll make a follow-up.

File: cobalt/cobalt.cc (hunk 16 above)
Comment (jellefoks) [reply cobalt/cobalt.cc:52]: Please clarify in the discussion thread what format to use and why.

File: cobalt/cobalt.cc (hunk 14 above)
Comment (jellefoks) [reply cobalt/cobalt.cc:44]: Maybe, but this way we're not dependent on upstream changing it from under us.

Note all this flags setting here should be removed later. This can't be the cleanest/right to do it. It's just the quickest to get it working ATM. See the todo mentioned above.

File: cobalt/cobalt.cc (hunk 10 above)
Comment (dahlstrom-g) [reply cobalt/cobalt.cc:48]: This thoughtful comment helped my update to 120.0.6099.331
(draft in commit 0ec0386d). Thanks!
